Userland/Libraries/LibC/netdb.cpp

```diff
@@ -895,11 +895,96 @@ int getaddrinfo(char const* __restrict node, char const* __restrict service, con
 
     long port;
     int socktype;
-    servent* svc_ent = nullptr;
-    if (!hints || (hints->ai_flags & AI_NUMERICSERV) == 0) {
-        svc_ent = getservbyname(service, proto);
+
+    struct ServiceData {
+        String protocol;
+        u32 port;
+    };
+    Optional<ServiceData> service_data = {};
+
+    if ((!hints || (hints->ai_flags & AI_NUMERICSERV) == 0) && service) {
+        services_file = fopen(services_path, "r");
+        if (!services_file) {
+            return EAI_FAIL;
+        }
+
+        while (true) {
+            char* line = nullptr;
+            size_t length = 0;
```

Comment:
Same comment here about looking more or less copied from the existing impl of getservbyname/fill_getserv_bufffers.

Can we consolidate the helpers, or refactor them, so that we can re-use the implementation here instead of having two copies of it? if GNU has a `getservbyname_r` extension, perhaps we could use that? (or invent it anyway...)

Reply:
GNU does have a `getservbyname_r` but implementing that is above my current abilities, but what I could do is refactor `fill_getserv_buffers` to return a struct with the data instead of directly populating the buffers. Opening and closing the service file would still exist in multiple places. Would that be satisfactory?

Reply:
yeah that sounds reasonable